Add information about userns-remap

mstanleyjones opened the pull request:
Add details about `userns-remap`

Fixes #1366

@estesp PTAL

mstanleyjones commented:
@pvnovarese PTAL

thaJeztah commented on engine/admin/index.md:
````diff
@@ -75,6 +75,133 @@ restart Docker. This method works for every Docker platform. The following
+    testuser:231072:65536
+    ```
+   
+    This means that user-namespaced processes started by `testuser` will be owne
````
s/owne/owned/

thaJeztah commented on engine/admin/index.md:
````diff
@@ -75,6 +75,133 @@ restart Docker. This method works for every Docker platform. The following
 }
 ```
 
+## Isolate containers within a user namespace
````
Wondering if;

- this topic should be in the "security" section
- it should be on its own page

If there's a dedicated page for this, we can copy some more of the technical information from https://docs.docker.com/engine/reference/commandline/dockerd/#daemon-user-namespace-options to the page, and remove it from the `dockerd` reference.

The reference docs contain quite some information that's important when using this feature (limitations, how to start a privileged container if userns is enabled, and the current limitation of having a single "global" mapping, instead of custom mappings per container)

If we migrate the content here, we can keep a short section in the daemon docs, but refer to the extended documentation elsewhere

thaJeztah commented on engine/admin/index.md:
```diff
@@ -75,6 +75,133 @@ restart Docker. This method works for every Docker platform. The following
+host.
+
+This re-mapping is relatively transparent to the container, but introduces some
+configuration complexity when the container needs access to privileged resources
```
we can remove "privileged" here; also files owned by "regular" users will be inaccessible.

The goal of user-namespaces is to give the user inside the container no access to _anything_ on the host (i.e., if the process _would_ be able to escape the container, it's basically harmless, as it won't be able to do anything).

mstanleyjones commented:
Addressed great feedback, split this off into its own topic and made some "What next" links in the configuration index, one of them pointing to the security landing page.

mstanleyjones commented:
Please review at https://deploy-preview-3678--docker-docs.netlify.com/engine/security/userns-remap/

pvnovarese commented on engine/security/userns-remap.md:
```diff
@@ -0,0 +1,249 @@
+UID on the host, which does not even map to a real user. This means the process
+has no privileges on the host system at all.
+
+> Multiple ranges
```
should we have an example here?  It's not clear what happens if you have multiple ranges, what the expected behavior should be, or if there is any other user action required to make use of them.

mstanleyjones commented on engine/security/userns-remap.md:
```diff
@@ -0,0 +1,249 @@
+> by adding multiple non-overlapping mappings for the same user or group in the
+> `/etc/subuid` or `/etc/subgid` file. In this case, Docker uses only the first
+> five mappings, in accordance with the kernel's limitation of only five entries
+> in `/proc/self/uid_map` and `/proc/self/gid_map`.
```
I need more info to work up these examples, I think. @estesp any thoughts?

mstanleyjones commented on engine/security/userns-remap.md:
```diff
@@ -0,0 +1,249 @@
+
+## Prerequisites
+
+1.  If you want to use an existing username or user ID, it must already exist.
```
OK, I just pushed an amendment that I think addresses this.

mstanleyjones commented on engine/security/userns-remap.md:
```diff
@@ -0,0 +1,249 @@
+
+It is very important that the ranges not overlap, so that a process cannot gain
+access in a different namespace. On most Linux distributions, system utilities
+manage the ranges for you when you add or remove users.
```
Talked to @pvnovarese offline and agreed that we can expand the docs on multiple ranges later. I filed #3704 for this.

estesp commented:
Thanks @mstanleyjones! We can chat about multiple ranges.. its a fairly advanced topic with probably a very small set of potential users.

fabiomartinelli commented:
I'm one of those potential users, it would be great to get more info and further tested examples about the multiple ranges topic.

mstanleyjones commented:
@fabiomartinelli can you tell us more about your use  case for using multiple ranges?
